### PR TITLE
Support Pico 4 (and probably Pico Neo 3)

### LINKED ISSUE #24: Pico Neo3 prebuilt APK
Hi Guys,
Would you consider provide prebuilt APK that is compatible with Pico devices? Currently there's only Huawei & Quest build in the release page.

### LINKED ISSUE #306: Any plans to support PICO 4?
Any plans to support PICO 4?

### PR DESCRIPTION
**Pull request by rsjtdrjgfuzkfg**
This PR introduces a new build variant "picoxr" using Pico's OpenXR SDK, and makes necessary changes to get Wolvic to run on the Pico 4. It is likely that the variant also works with at least some versions of the Pico Neo 3, but that was not tested.

Official 3D models of the Pico 4 controllers are available as part of the Pico SDKs, but it is unclear if they are license-compatible with this repository. They are thus not included in this PR.

Fixes #306
Fixes #24 (untested, but Pico Neo 3 is documented to use the same SDK than the Pico 4)

### CODE REVIEW

**Comment by rsjtdrjgfuzkfg** [comment]
Test build: https://github.com/rsjtdrjgfuzkfg/wolvic/releases/tag/testbuild-pico-2022-11-27

If somebody knows of an easy way to enable WebXR, feel free to post ideas or amend this  PR.

File: README.md
Comment (svillar) [inline README.md:42]: I'll comment about the third_party stuff later, but we'll have to change this clarification)

File: app/CMakeLists.txt
Comment (svillar) [inline app/CMakeLists.txt:121]: We don't need to strictly adhere to the packaging of the SDK. In the third party repo we usually have a folder per platform and then just the minimum required files for the build. This means that it's enough to have
```
picooxr/Include
picooxr/Libs
```

File: app/CMakeLists.txt
Comment (svillar) [inline app/CMakeLists.txt:126]: Ditto about the paths. It should be enough to have `Libs` under `picooxr`

File: app/CMakeLists.txt
Comment (svillar) [inline app/CMakeLists.txt:208]: Ditto.

File: app/src/main/cpp/native-lib.cpp
Comment (svillar) [inline app/src/main/cpp/native-lib.cpp:194]: This is interesting, is there any bug open in a public repository about this. If that's the case we should include a link here.

File: app/src/openxr/cpp/DeviceDelegateOpenXR.cpp
Comment (svillar) [inline app/src/openxr/cpp/DeviceDelegateOpenXR.cpp:232]: This is interesting, however it should only apply to Pico4 devices. Note that this new flavour should be usable also for devices with different resolutions.

File: app/src/openxr/cpp/DeviceDelegateOpenXR.cpp
Comment (svillar) [inline app/src/openxr/cpp/DeviceDelegateOpenXR.cpp:237]: You could do 
`viewConfig.front().recommendedImageRectWidth = std::min(viewConfig.front().maxImageRectWidth, viewConfig.front().recommendedImageRectWidth);`

Same for the max size

File: app/src/openxr/cpp/OpenXRInputMappings.h
Comment (svillar) [inline app/src/openxr/cpp/OpenXRInputMappings.h:196]: Haven't seen any interaction profile path published by Khronos so far, I guess you're simply making it up?

**Review by svillar — CHANGES_REQUESTED** [review]
First of all, many thanks for your contribution. It's really appreciated. I've added some minor things to change here in the review. Apart from that I have a couple of comments/requests:
1. Replace picooxr by picoxr. I understand the motivation of your chosen name, but the double 'o' is hard to read and I'm sure will lead to developer frustration when incorrectly typing that name
2. There are a couple of things related to controllers that you mention but I was unable to find. For example the OpenXR interaction profile and the 3D models of the controllers. I'd love to have pointers to those assests

**Comment by svillar** [comment]
> Test build: https://github.com/rsjtdrjgfuzkfg/wolvic/releases/tag/testbuild-pico-2022-11-27 (contains controller model not included in this PR)
> 
> If somebody knows of an easy way to enable WebXR, feel free to post ideas or amend this PR.

BTW WebXR does work fine with your patch. If it's not working for you it's likely because you are not using the correct  `gecko-dev` branch. See the [setup instructions](https://github.com/Igalia/wolvic/#setup-instructions) for additional details

**Comment by svillar** [comment]
I've already uploaded the SDK to the third_party private repository. As mentioned in the review I placed just the `Libs` and `Include` dirs directly under `picoxr` repository. The rest of the SDK is not really needed.

BTW @rsjtdrjgfuzkfg I didn't mention it in the review, but if you don't plan to do further work on the PR just tell us and we'll complete the missing bits. Thanks again.

File: app/src/common/shared/com/igalia/wolvic/utils/DeviceType.java
Comment (svillar) [inline app/src/common/shared/com/igalia/wolvic/utils/DeviceType.java:59]: "Pico OXR" -> "Pico XR"

Also do please add a new isPicoXR() method by the very end of this file as well.

**Review by svillar — CHANGES_REQUESTED** [review]

File: app/CMakeLists.txt
Comment (rsjtdrjgfuzkfg) [reply app/CMakeLists.txt:121]: I'd prefer to keep the original folder structure, as that makes it easier for people without access to your private repository to build the project. This is independent of not adding unnecessary files to your private repository: you can just omit them when setting up your repository.

That said, in this specific case, I could live with dropping the `OpenXR` folder if it is a big deal to you, as it seems you keep the structure identical (including the uncommon capitalization made by Pico).

File: app/src/main/cpp/native-lib.cpp
Comment (rsjtdrjgfuzkfg) [reply app/src/main/cpp/native-lib.cpp:194]: I found this issue while debugging and only saw it in Wolvic, so I don't think there is a mention anywhere. It is also unlikely to be a general issue, as I did not see this with the Godot game engine. I'd guess that the callback is not yet registered when the init and resume events happen, possibly during the setup of Pico's VR runtime bits that "upgrade" a normal android activity. Whether or not this is caused by Wolvic or Pico is not something I can say for certain, though.

File: app/src/openxr/cpp/DeviceDelegateOpenXR.cpp
Comment (rsjtdrjgfuzkfg) [reply app/src/openxr/cpp/DeviceDelegateOpenXR.cpp:232]: It should work on all devices, but of course waste performance on devices that do not benefit from the higher resolution.

If you know of a certain way to identify the Pico 4, feel free to suggest something? The main reason I specified a factor instead of a hardcoded resolution was that I hoped that the Neo 3 recommends a smaller resolution. (The other was that I think it would be nice to make that factor a setting for users in the long term, on all headsets.)

File: app/src/openxr/cpp/OpenXRInputMappings.h
Comment (rsjtdrjgfuzkfg) [reply app/src/openxr/cpp/OpenXRInputMappings.h:196]: I used information I got from the Pico developers that assist over at Godot, so not totally made up.

That said, Pico is indeed behind on publishing things through Khronos, they claim to be working on amending the OpenXR spec with their controllers and extensions, but I do not know when that will actually happen / get published.

**Comment by rsjtdrjgfuzkfg** [comment]
@svillar thanks for clarifying the gecko override is still necessary; I did follow some links and read that it is no longer necessary – but it seems I misunderstood something.

Regarding the controller models: the Pico SDK for Unity contains 3D models for all Pico controllers, likely other Pico SDKs do as well. They are not ready to use for our purposes, though. The build I uploaded over at my repo contains a manual conversion of the Pico 4 controllers to the right format, but the SDK license is unclear about redistribution outside of binary software releases.

I did send a mail out to the Pico developer that helped with licensing issues over at Godot, so I hope Pico permits us to use the models under MPL2.0 or finds us some other solution to keep them in this repository.

Regarding the naming: picoxr is also a term Pico seems to use for pretty much everything, so I wanted to avoid it. But I admit I also had some difficulty in typing it consistently, so I'm fine with removing the second `o`. ;)

**Comment by rsjtdrjgfuzkfg** [comment]
I updated the PR and the test build (second comment). I can confirm WebXR is now working for me as well.

From my POV, there are two remaining issues: the controller models for the Pico 4 and better support for Pico XR devices other than the Pico 4. The latter is not something I can help with, and I think that could go into a separate PR after this one is merged.

Regarding the former: I haven't heard back from Pico regarding controller model licensing yet, but will update you when I hear something. I'll leave it up to you if you want to delay merging this PR for that.

**Review by svillar — APPROVED** [review]
I'm fine merging this now even without the controllers' models and pictures for the interstitial.

Thank you very much!

File: app/src/openxr/cpp/DeviceDelegateOpenXR.cpp
Comment (rpavlik) [reply app/src/openxr/cpp/DeviceDelegateOpenXR.cpp]: This honestly seems like a runtime bug that shouldn't be worked around in applications

**Comment by svillar** [comment]
BTW @rsjtdrjgfuzkfg mind removing the package you have in https://github.com/rsjtdrjgfuzkfg/wolvic/releases/tag/testbuild-pico-2022-11-30 ? I know your intentions were good but it looks like people is trying it as it was an official release build, so they're reporting bugs when the support is still in very early stages. I think it does not provide a good image of the project to have a half-backed release. Thanks

**Comment by rsjtdrjgfuzkfg** [comment]
@svillar given that you (as an organization) own the branding, I will comply with your request. Can't say I like it though. Any timeframe for you building Pico officially and/or permitting third parties to release branded builds?

Adding a bit of build logic to permit unbranded test builds might be a good idea to permit people not affiliated with Igala to freely release test builds. If the unbranded version has some clear disclaimers in-app, that should also solve the "people reporting issues for non-official releases" thing.

**Comment by svillar** [comment]
> @svillar given that you (as an organization) own the branding, I will comply with your request. Can't say I like it though. Any timeframe for you building Pico officially and/or permitting third parties to release branded builds?
> 
> Adding a bit of build logic to permit unbranded test builds might be a good idea to permit people not affiliated with Igala to freely release test builds. If the unbranded version has some clear disclaimers in-app, that should also solve the "people reporting issues for non-official releases" thing.

Well don't get me wrong, Wolvic is free software and we love that with all those implications. It was just a suggestion mainly driven by the huge amount of tasks we have to carry on ATM. Taking care of reports from non-released builds will just pile up more over over our shoulders.

File: app/src/openxr/cpp/DeviceDelegateOpenXR.cpp
Comment (svillar) [reply app/src/openxr/cpp/DeviceDelegateOpenXR.cpp]: @rpavlik is Pico using Monado? If that's the case I have another one to report :smile:

Comment (rpavlik) [reply app/src/openxr/cpp/DeviceDelegateOpenXR.cpp]: Not unmodified at least.

**Comment by ipeakermit** [comment]
@rsjtdrjgfuzkfg - do you have a moment to help me replicate your WebXR build for picoxr / Pico 3? I'm having trouble building. See https://github.com/Igalia/wolvic/issues/416

**Pull request merged**
